rfc/20240824-provider-cache-locking.md

```diff
@@ -0,0 +1,68 @@
+# Safe Global Provider Cache Access
+
+Issue: https://github.com/OpenTofu/opentofu/issues/1483
+
+Many CI/CD systems download all providers for every single tofu action run. The global provider cache allows downloaded providers to be saved in between runs and linked into the local provider cache (in .terraform/). Unfortunately, the global provider cache does not have any locking around it and can fail when used in multiple simultaneous runs.
+
+Another common use case is terragrunt. When running `tofu init` on multiple projects simultaneously via the tool, there is a high likely hood of a conflicting write to a provider's directory. This has caused some interesting tooling workarounds on their end, which are not ideal.
+
+Additionally, as we build out true e2e tests for OpenTofu, safe access to the global provider cache can dramatically reduce run times.
+
+## Proposed Solution
+
+A filesystem level lock should be added that is safe for cross-process and in some scenarios cross-machine access. It will be best-effort and should rely on standard locking practices (not home grown) and be transparent to the user.
```

Comment:
I'm a bit worried that this can be mis-understood. A "filesystem level lock" could also just be a lockfile written to disk.

We should mention that we're looking at using the native syscalls to lock files at the filesystem level here. 

Especially when we talk about lockfiles below. I don't want people to mix up the 2

Reply:
ebf7dae4c4